File: Placement Sheet/Arrays/Initial Startups/Pair_Sum.java
import java.util.HashMap;
import java.util.Scanner;

public class Pair_Sum {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        System.out.print("Enter the size of the array: ");
        int size = input.nextInt();

        int[] arr = new int[size];
        for (int i = 0; i < size; i++) {
            System.out.printf("Enter %d number element of the array: ", i + 1);
            arr[i] = input.nextInt();
        }

        System.out.print("Enter the target value: ");
        int target = input.nextInt();
        input.close();
        System.out.println("Pair sum: " + pairSumBruteForce(arr, size, target));
        System.out.println("Pair sum: " + pairSum(arr, size, target));
    }

    // Brute-Force
    private static int pairSumBruteForce(int[] arr, int n, int target) {
        if (n == 1) {
            return -1;
        }
        int count = 0;
        for (int i = 0; i < n - 1; i++) {
            for (int j = i + 1; j < n; j++) {
                if (arr[i] + arr[j] == target) {
                    count++;
                }
            }
        }
        return count == 0 ? -1 : count;
    }

    // Optimal using hashing
    private static int pairSum(int[] arr, int n, int target) {

        if (n == 0) {
            return -1;
        }

        HashMap<Integer, Integer> map = new HashMap<>();

        int count = 0;
        for (int i = 0; i < n; i++) {
            int complement = target - arr[i];
            if (map.containsKey(complement)) {
                count += map.get(complement);
            }
            map.put(arr[i], map.getOrDefault(arr[i], 0) + 1);
        }

        return count == 0 ? -1 : count;
    }
}

File: Placement Sheet/Arrays/Initial Startups/Python/Pair_Sum.py
# Brute-Force
def pairSumBruteForce(arr, target):
    if arr is None or len(arr) == 0:
        return -1

    count = 0
    for i in range(0, len(arr) - 1):
        for j in range(i + 1, len(arr)):
            if arr[i] + arr [j] == target:
                count += 1
    if count == 0:
        return -1
    else:
        return count

# Optimal...
def pairSum(arr, target):
    if arr is None or len(arr) == 0:
        return -1

    map = {}
    count = 0

    for i in range(0, len(arr)):
        complement = target - arr[i]
        if complement in map:
            count += map[complement]
        map[arr[i]] = map.get(arr[i], 0) + 1

    return -1 if count == 0 else count

size = int(input("Enter the size of the array: "))
arr = []
for i in range(0, size):
    data = int(input("Enter the data: "))
    arr.append(data)

target = int(input("Enter the target element: "))

result = pairSumBruteForce(arr=arr, target=target)
